src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,548 @@
+    /// - faulty_blocks.count() = 0 implies faulty_blocks_repair_index = faulty_blocks.count()
+    faulty_blocks_repair_index: usize = 0,
+
+    /// Invarants:
+    /// - enqueued_table_blocks + enqueued_single_blocks = faulty_blocks.count()
+    /// - enqueued_table_blocks ≤ options.tables_max * lsm_table_content_blocks_max
+    enqueued_single_blocks: usize = 0,
+    enqueued_table_blocks: usize = 0,
+
+    /// Invariants:
+    /// - For every index address in faulty_tables: ¬free_set.is_free(address).
+    faulty_tables: FIFO(RepairTable) = .{ .name = "grid_repair_queue_tables" },
+
+    checkpointing: ?struct {
+        /// The number of faulty_blocks with state=aborting.
+        aborting: usize,
+    } = null,
+
+    /// Guard against new block/table repairs queueing within RepairTable callbacks.
+    canceling: bool = false,
+
+    pub fn init(allocator: std.mem.Allocator, options: Options) error{OutOfMemory}!GridRepairQueue {
+        var faulty_blocks = FaultyBlocks{};
+        errdefer faulty_blocks.deinit(allocator);
+
+        try faulty_blocks.ensureTotalCapacity(
+            allocator,
+            options.blocks_max + options.tables_max * constants.lsm_table_content_blocks_max,
+        );
+
+        return GridRepairQueue{
+            .options = options,
+            .faulty_blocks = faulty_blocks,
+        };
+    }
+
+    pub fn deinit(queue: *GridRepairQueue, allocator: std.mem.Allocator) void {
+        assert(!queue.canceling);
+
+        queue.faulty_blocks.deinit(allocator);
+
+        queue.* = undefined;
+    }
+
+    /// When the queue wants more blocks than fit in a single request message, successive calls
+    /// to this function cycle through the pending BlockRequests.
+    pub fn next_batch_of_block_requests(
+        queue: *GridRepairQueue,
+        requests: []vsr.BlockRequest,
+    ) usize {
+        assert(!queue.canceling);
+        maybe(requests.len == 0);
+
+        const faults_total = queue.faulty_blocks.count();
+        if (faults_total == 0) return 0;
+        assert(faults_total > queue.faulty_blocks_repair_index);
+
+        const fault_addresses = queue.faulty_blocks.entries.items(.key);
+        const fault_data = queue.faulty_blocks.entries.items(.value);
+
+        var requests_count: usize = 0;
+        var fault_offset: usize = 0;
+        while (requests_count < requests.len and
+            fault_offset < faults_total) : (fault_offset += 1)
+        {
+            const fault_index =
+                (queue.faulty_blocks_repair_index + fault_offset) % faults_total;
+
+            switch (fault_data[fault_index].state) {
+                .waiting => {
+                    requests[requests_count] = .{
+                        .block_address = fault_addresses[fault_index],
+                        .block_checksum = fault_data[fault_index].checksum,
+                    };
+                    requests_count += 1;
+
+                    if (requests_count == requests.len) break;
```

Comment:
I think this `break` is fully redundant with `requests_count < requests.len` while condition, so we  could remove either the break, or the extra condition in while (I'd prune the while)